Add the Thrift tutorial (#4865)
line/armeria: site/src/pages/tutorials/thrift/blog/run-server-and-client.mdx

---
menuTitle: "Run a server and client"
order: 2
category: thrift
type: step
targetLang: java
---

# Running a server and client

In this step, we'll create and run a server with the service we created in the previous step.
Also, we'll create a client and make a call to the service with the client.

<TutorialSteps current={2} />

## What you need

No preparation is required for this step.

## 1. Create a server

Let's create a server to serve our service.

1. Create the `Main` class for the server.
  You can see the full version of the file [here](https://github.com/line/armeria-examples/tree/main/tutorials/thrift/src/main/java/example/armeria/server/blog/thrift/Main.java).
  ```java filename=Main.java
  package example.armeria.server.blog.thrift;

  import org.slf4j.Logger;
  import org.slf4j.LoggerFactory;

  public final class Main {
      private static final Logger logger = LoggerFactory.getLogger(Main.class);
  }
  ```
2. Add the `newServer()` method in the `Main` class as follows.

>   A service instance is created using Armeria's <type://THttpService#builder()> in the method.

Author: How about changing the sentence instead of adding a tip?

```suggestion
  We are using Armeria's <type://THttpService> to handle Thrift calls.
```

Generally speaking, it's not a good idea to include too much conceptual information to a tutorial.

>   A service instance is created using Armeria's <type://THttpService#builder()> in the method.

Reviewer: > Generally speaking, it's not a good idea to include too much conceptual information to a tutorial.

I wasn't aware of this 😅    I can agree with that.

> How about changing the sentence instead of adding a tip?

Sounds good to me 👍

  ...
  private static Server newServer(int port) throws Exception {
    ...
    return Server.builder()
                 .http(port)
                 .service("/thrift", tHttpService) // Add the service to server
                 .build();
  }
  ```
4. Add the `main()` method in the `Main` class as follows.
  ```java filename=Main.java
  public static void main(String[] args) throws Exception {
    final Server server = newServer(8080);

    server.closeOnJvmShutdown().thenRun(() -> {
      logger.info("Server has been stopped.");
    });

    server.start().join();
  }
  ```

## 2. Run the server

Run the `Main.main()` method on your IDE or using Gradle.
```bash
./gradlew run
```

Your server is running if you see the following message.
```bash
[armeria-boss-http-*:8080] INFO com.linecorp.armeria.server.Server - Serving HTTP at /[0:0:0:0:0:0:0:0]:8080 - http://127.0.0.1:8080/
```

## 4. Create a client

1. Create a main class for our client. We'll name the class `BlogClient`.
  ```java filename=BlogClient.java
  package example.armeria.server.blog.thrift;

  import org.slf4j.Logger;
  import org.slf4j.LoggerFactory;

  public class BlogClient {

>     private static final Logger logger = LoggerFactory.getLogger(BlogClient.class);

Author: It's used in the `main()` method in the `BlogClient`  class, and `5. Run the client` step on the same page.

>     private static final Logger logger = LoggerFactory.getLogger(BlogClient.class);

Reviewer: I see 😅  nevermind this comment then 🙇

  }
  ```
2. In the `BlogClient` class, add a constructor and create a Thrift client instance using Armeria's <type://ThriftClients>.
  ```java filename=BlogClient.java
  import com.linecorp.armeria.client.thrift.ThriftClients;
  import example.armeria.blog.thrift.BlogService;

  public class BlogClient {
    ...
    private final BlogService.Iface blogService;

    BlogClient(URI uri, String path) {
      blogService = ThriftClients.builder(uri)
                .path(path)
                .build(BlogService.Iface.class);
    }
  }
  ```
3. Add the `hello()` method to make a request to the `hello()` method of our service.
  ```java filename=BlogClient.java
    String hello() throws TException {
      return blogService.hello();
    }
  ```

> 4. Add the `main()` method as follows.

Author: It sounds good to me! I agree that it's sensible to avoid having two `main()` methods.
Will do 🙇‍♀️

> 4. Add the `main()` method as follows.

Author: Done: 8e1ec88
Please take a look and feel free to give me further opinions!
